File: code/com/leetcode/solutions/NO_0019_删除链表的倒数第N个节点/Solution.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * 使用两个间隔为n的指针，当后面的指针移动到结尾是，前面的指针即为删除位
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
    if (0 === n) { return head.next };
    let p = head, q = head;
    while (n--) { q = q.next };

    if (null === q) { return head.next; }

    while (q.next) {
        p = p.next;
        q = q.next;
    }
    p.next = p.next.next;
    return head;
};

File: code/com/leetcode/solutions/NO_0019_删除链表的倒数第N个节点/Solution.cs
using System;

public class Solution
{

    /*
    * 思路：定义LinkedListNode 找到第Length -1 的位置，移除节点后修改前后节点的上一个节点的next 和下一个节点的prev
    * 其他：
   */

    public LinkedListNode<int> removeNthFromEnd(LinkedListNode<int> head, int n)
    {

        LinkedListNode<int> dummy = new LinkedListNode<int>(0);
        dummy = head;

        LinkedListNode<int> rear = dummy;
        LinkedListNode<int> pre = dummy;
        int a = 0;
        while (pre.Next != null)
        {
            a++;
            pre = pre.Next;
            if (a >= n + 1)
            {

                rear = rear.Next;
            }
        }
        rear = rear.Next.Next;
        return dummy.Next;
    }

}
